rex/external/rexarray.py

```diff
@@ -0,0 +1,1037 @@
+"""
+import io
+import os
+import json
+import logging
+import warnings
+from pathlib import Path
+from collections import namedtuple
+
+import h5py
+import numpy as np
+import dask.array as da
+from xarray import coding
+from xarray.backends.common import (AbstractDataStore, BackendArray,
+                                    BackendEntrypoint, _normalize_path)
+from xarray.backends.file_manager import CachingFileManager
+from xarray.backends.locks import (HDF5_LOCK, combine_locks, ensure_lock,
+                                   get_write_lock)
+from xarray import conventions
+from xarray.core import indexing
+from xarray.core.dataset import Dataset
+from xarray.core.utils import (FrozenDict, is_remote_uri,
+                               read_magic_number_from_file,
+                               try_read_magic_number_from_file_or_path,
+                               close_on_error)
+from xarray.core.variable import Variable
+
+from rex.utilities import (rex_unscale, import_fsspec_or_fail, is_hsds_file,
+                           import_h5pyd_or_fail, assert_read_only_mode)
+
+
+logger = logging.getLogger(__name__)
+TI_DTYPE = np.dtype('datetime64[ns]')
+_SA_FN = Path(__file__).parent / "standard_attrs.json"
+_EN_FN = Path(__file__).parent / "encondings.json"
+_SA = {}
+_EN = {}
+
+
+VarInfo = namedtuple("VarInfo", ["name", "var", "meta_index", "coord_index"],
+                     defaults=[-1, -1])
+
+
+class RexMetaVar:
+    """Wrapper class containing meta attributes for a variable"""
+
+    __slots__ = ("chunks", "fletcher32", "shuffle", "dtype", "shape",
+                 "compression", "compression_opts", "attrs")
+
+    def __init__(self, meta_var, dtype):
+        """
+
+        Parameters
+        ----------
+        meta_var : handle
+            Handle that contains meta attributes for the coordinate.
+        dtype : str | obj
+            Variable data type.
+        """
+        self.chunks = meta_var.chunks
+        self.fletcher32 = meta_var.fletcher32
+        self.shuffle = meta_var.shuffle
+        self.dtype = dtype
+        self.shape = meta_var.shape
+        self.compression = None
+        self.compression_opts = None
+        self.attrs = {}
+
+
+class RexCoordVar:
+    """Wrapper class containing meta attributes for a coordinate"""
+
+    __slots__ = ("chunks", "fletcher32", "shuffle", "dtype", "shape",
+                 "compression", "compression_opts", "attrs")
+
+    def __init__(self, coord_var):
+        """
+
+        Parameters
+        ----------
+        coord_var : handle
+            Handle that contains meta attributes for the coordinate.
+        """
+        self.chunks = coord_var.chunks
+        self.fletcher32 = coord_var.fletcher32
+        self.shuffle = coord_var.shuffle
+        self.dtype = coord_var.dtype
+        self.shape = coord_var.shape[:-1]
+        self.compression = None
+        self.compression_opts = None
+        self.attrs = {}
+
+
+class RexArrayWrapper(BackendArray):
```

Comment:
This file mostly follows xarray code, right? Could you point to the main differences I should focus on? Is the main new thing this wrapper class?

Reply:
Yup, this class and `RexStore` are the ones that contain most of the rex-specific changes. I would focus on those two